### PR TITLE
8296924: C2: assert(is_valid_AArch64_address(dest.target())) failed: bad address

### PR DESCRIPTION
**Pull request by TobiHartmann**
With (unreachable) unsafe accesses, it can happen that the base address is invalid. On AArch64, C2 will emit a `loadConP` for loading the constant address that is implemented by [aarch64_enc_mov_p](https://github.com/openjdk/jdk/blob/48017b1d9c3a7867984f54d61f17c7f034d213f5/src/hotspot/cpu/aarch64/aarch64.ad#L3366) calling [MacroAssembler::adrp](https://github.com/openjdk/jdk/blob/48017b1d9c3a7867984f54d61f17c7f034d213f5/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L4576). The `adrp` implementation then asserts in [is_valid_AArch64_address](https://github.com/openjdk/jdk/blob/48017b1d9c3a7867984f54d61f17c7f034d213f5/src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp#L1321), assuming that we can only ever load constant pointers that are within the 48-bit AArch64 address space.

The fix, proposed by @theRealAph, is to emit a full-blown `mov` in case of a bad address.

Thanks,
Tobias

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296924](https://bugs.openjdk.org/browse/JDK-8296924): C2: assert(is_valid_AArch64_address(dest.target())) failed: bad address

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to [fa945fa3](https://git.openjdk.org/jdk/pull/11412/files/fa945fa333b91aa35bc05bad8cc5c5dbde9c3292)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Contributors
 * Andrew Haley `<aph@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11412/head:pull/11412` \
`$ git checkout pull/11412`

Update a local copy of the PR: \
`$ git checkout pull/11412` \
`$ git pull https://git.openjdk.org/jdk pull/11412/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11412`

View PR using the GUI difftool: \
`$ git pr show -t 11412`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11412.diff">https://git.openjdk.org/jdk/pull/11412.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/aarch64.ad [hunk 1]
```diff
@@ -3377,7 +3377,8 @@ encode %{
         __ mov_metadata(dst_reg, (Metadata*)con);
       } else {
         assert(rtype == relocInfo::none, "unexpected reloc type");
-        if (con < (address)(uintptr_t)os::vm_page_size()) {
+        if (!__ is_valid_AArch64_address(con) ||
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64.ad:3380]:
```suggestion
        if (! __ is_valid_AArch64_address(con) ||
```

**Review by theRealAph — APPROVED** [review]
Thanks. That's obviously correct.

**Comment by TobiHartmann** [comment]
Thanks for the review and the help with the fix, Andrew!

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by TobiHartmann** [comment]
Thanks for the review, Vladimir!

**Pull request closed without merging**